# Constrative-sc

This repository contains the pytorch implementation of the paper "Contrastive sel-supervised clustering of scRNA-seq data", by Madalina Ciortan under the supervision of Matthieu Defrance.  

In this work, we propose contrastive-sc, an unsupervised deep learning method to cluster scRNA-seq data using contrastive representation learning in a two phased process. First, self-supervised representation learning is used on strongly augmented cell data to learn an embedding which is clustered in a second phase. Several strategies for clustering have been explored, ranging from (a) clustering directly the resulting embedding with an arbitrary algorithm (e.g. Leiden community detection, K-Means) , (b) enriching the network with a clustering layer and a soft k-Means loss or by (c) creating a combined representation of multiple estimators (an ensemble) and clustering this embedding with an arbitrary algorithm. Multiple experiments have been conducted on simulated data (both class balanced and class imbalanced) and on real sequenced data. The ensemble representation combined with Leiden community detection (c) performed best and surpassed state-of-the-art competitors when the number of clusters or the dropout rate are increasing. 

# Overview of the repository
- **notebooks** folder contains all jupyter notebooks to run the project, as detailed below.
- **others** folder contains the code to reproduce all experiments with scanpy, sczi, scDeepCluster
- **R** folder contains the scrips to generate the simulated data in folder R/simulated_data (both balanced and imbalanced)
- **outoput** contains model dumps and the results of running all experiments, needed to reproduce the plots
- **docker** contains the Dockerfile to create the image used to run all python experiments
- **real_data** contains the biological scRNA-seq data, downloaded from scDeepCluster, as detailed below
- train.py contains the main functionalities for training and evaluating the model results
- model.py contains the network definition
- st_loss.py contains the implementation of the loss functions
- utils.py contains various utility functions
- loop.py contains the same code as train.py but at a lower granularity

### Overview of notebooks
- **Main.ipynb** represents the main entry points, contains code snipped to train the model on the simulated or real data
- **Benchmark real data, simulated balalanced and simulated imbalanced** contains the code to reproduce all experiments
- **Plot simulated balanced, simulated imbalanced, real data and combined** contains code to reproduce all plots
- **Input size analysis and Dropout analysis** explore the importance of the number of selected genes and that of the data augmentation

## Environment Setup
We have employed a docker containers to facilitate reproducing the paper results.

### Python environment
It can be launched by running the following:

cd docker
docker build -t contrastive-sc .

The image has been created for GPU usage. In order to run it on CPU, in the Dockerfile, the line "pytorch/pytorch:1.4-cuda10.1-cudnn7-runtime" should be replaced with a CPU version.

The command above created a docker container tagged as contrastive-sc . Assuming the project has been cloned locally in a parent folder named notebooks, the image can be launched locally with:

docker run -it --runtime=nvidia -v ~/notebooks:/tf/notebooks -p 8888:8888 contrastive-sc

This starts up a jupyter notebook server, which can be accessed at http://localhost:8888/tree/notebooks

### R environment

We followed the instructions on this [tutorial](http://bioinformatics.sph.harvard.edu/knowledgebase/scrnaseq/rstudio_sc_docker.html) in order to create an R docker container which comes with most single-cell related libraries already installed.
In order to launch it on port 8787, execute the following:


docker run -d -p 8787:8787 -e USER='rstudio' -e PASSWORD='rstudioSC' -e ROOT=TRUE -v ~/notebooks/deep_clustering:/home/rstudio/projects vbarrerab/rstudio_singlecell



## Data
The simulated datasets can be downloaded from this [Google Drive link](https://drive.google.com/file/d/18OMv_uTQs8WespsPGc-g_kTd1PXxpmpH/view?usp=sharing) (~9GB). Alternatively, it can be generated by running R/all_balanced.r or R/all_imbalanced.R.  

The single cell data has been published and can be accessed from the 
[scDeepCluster repository](https://github.com/ttgump/scDeepCluster). It should be saved to **real_data** folder.